dex/governance/src/contract.rs

```diff
@@ -0,0 +1,138 @@
+
+        let vote_weight = self.get_vote_weight(&payment);
+        let min_weight = self.min_weight_for_proposal().get();
+        require!(vote_weight >= min_weight, NOT_ENOUGH_FUNDS_TO_PROPOSE);
+
+        let mut proposal = self.new_proposal_from_args(args);
+        self.proposal_id_counter().set(proposal.id + 1);
+
+        proposal.num_upvotes = vote_weight.clone();
+        self.proposal(proposal.id).set(&proposal);
+
+        let vote_nft = self.create_vote_nft(proposal.id, VoteType::Upvote, vote_weight, payment);
+        self.send_back(vote_nft);
+
+        proposal.id
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn upvote(&self, proposal_id: u64) {
+        self.vote(proposal_id, VoteType::Upvote)
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn downvote(&self, proposal_id: u64) {
+        self.vote(proposal_id, VoteType::DownVote)
+    }
+
+    #[endpoint]
+    fn execute(&self, proposal_id: u64) {
+        require!(!self.proposal(proposal_id).is_empty(), PROPOSAL_NOT_FOUND);
+        let mut proposal = self.proposal(proposal_id).get();
+
+        let pstat = self.get_proposal_status(&proposal);
+        require!(pstat == ProposalStatus::Succeeded, PROPOSAL_NOT_SUCCEEDED);
+
+        self.execute_proposal(&proposal);
+        proposal.was_executed = true;
+        self.proposal(proposal_id).set(&proposal);
+    }
+
+    fn vote(&self, proposal_id: u64, vote_type: VoteType) {
+        require!(!self.proposal(proposal_id).is_empty(), PROPOSAL_NOT_FOUND);
+        let mut proposal = self.proposal(proposal_id).get();
+
+        let pstat = self.get_proposal_status(&proposal);
+        require!(pstat == ProposalStatus::Active, PROPOSAL_NOT_ACTIVE);
+
+        let payment = self.call_value().payment();
+        self.require_is_accepted_payment_for_voting(&payment);
+
+        let vote_weight = self.get_vote_weight(&payment);
+        match vote_type {
+            VoteType::Upvote => proposal.num_upvotes += &vote_weight,
+            VoteType::DownVote => proposal.num_downvotes += &vote_weight,
+        }
+
+        let vote_nft = self.create_vote_nft(proposal.id, vote_type, vote_weight, payment);
+        self.send_back(vote_nft);
+
+        self.proposal(proposal_id).set(&proposal);
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn redeem(&self) {
+        let payment = self.call_value().payment();
+
+        let vote_nft_id = self.vote_nft_id().get();
+        require!(payment.token_identifier == vote_nft_id, BAD_TOKEN_ID);
+
+        let attr = self.get_vote_attr(&payment);
+        let proposal = self.proposal(attr.proposal_id).get();
+
+        let pstat = self.get_proposal_status(&proposal);
+        match pstat {
+            ProposalStatus::Defeated | ProposalStatus::Executed => {
+                self.send_back(attr.payment);
+                self.burn_vote_nft(payment);
+            }
+            ProposalStatus::Succeeded => {
+                sc_panic!(PROPOSAL_NEEDS_TO_BE_EXECUTED)
```

Comment:
I think tokens can be redeemed back if it was  already voted - no need for execution. But I see your point - however, people could block tokens in some proposals which will never be executed as someone veto's is.